packages/ts/hilla-file-router/src/vite-plugin-file-router.ts

```diff
@@ -0,0 +1,109 @@
+import { writeFile } from 'node:fs/promises';
+import { fileURLToPath, pathToFileURL } from 'node:url';
+import type { Plugin } from 'vite';
+import collectRoutes from './collectRoutes.js';
+import generateJson from './generateJson.js';
+import generateRoutes from './generateRoutes.js';
+
+export type PluginOptions = Readonly<{
+  /**
+   * The base directory for the router. The folders and files in this directory
+   * will be used as route paths.
+   *
+   * @defaultValue `frontend/views`
+   */
+  viewsDir?: URL | string;
+  /**
+   * The directory where the generated view file will be stored.
+   *
+   * @defaultValue `frontend/generated`
+   */
+  generatedDir?: URL | string;
+  /**
+   * The list of extensions that will be collected as routes of the file-based
+   * router.
+   *
+   * @defaultValue `['.tsx', '.jsx', '.ts', '.js']`
+   */
+  extensions?: readonly string[];
+  /**
+   * The name of the export that will be used for the {@link ViewConfig} in the
+   * route file.
+   *
+   * @defaultValue `config`
+   */
+  configExportName?: string;
+}>;
+
+type GeneratedUrls = Readonly<{
+  json: URL;
+  code: URL;
+}>;
+
+async function generate(code: string, json: string, urls: GeneratedUrls) {
```

Comment:
If this function creates some files, I suggest renaming it to `createXXXfile(s)` or simply `writeToFile`, as `generate` is misleading.